8267189: Remove duplicated unregistered classes from dynamic archive

xiangzhai opened the pull request:
Hi,

Backporting JDK-8267189 as a prerequisite for JDK-8275731 (CDS archived enums objects are recreated at runtime)

The patch applies clean except manually edit the source file:

* src/hotspot/share/classfile/systemDictionaryShared.hpp
Made `static void SystemDictionaryShared::remove_dumptime_info(InstanceKlass*)` public within this context.

Tested jtreg tier1~3, no regression.

Thanks,
Leslie Zhai

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8267189](https://bugs.openjdk.org/browse/JDK-8267189): Remove duplicated unregistered classes from dynamic archive


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk17u-dev pull/939/head:pull/939` \
`$ git checkout pull/939`

Update a local copy of the PR: \
`$ git checkout pull/939` \
`$ git pull https://git.openjdk.org/jdk17u-dev pull/939/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 939`

View PR using the GUI difftool: \
`$ git pr show -t 939`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk17u-dev/pull/939.diff">https://git.openjdk.org/jdk17u-dev/pull/939.diff</a>

</details>

xiangzhai commented:
/integrate

iklam commented:
> Hi @iklam
> 
> Could you sponsor it?
> 
> Thanks, Leslie Zhai

The bug still has the "jdk17u-fix-request" label. It needs to be approved by jdk17u maintainer first.

iklam commented:
The description of this PR is not clear. Are you backporting JDK-8267189 because it's required by [JDK-8275731](https://bugs.openjdk.org/browse/JDK-8275731) "CDS archived enums objects are recreated at runtime", which is the ultimate issue that you want to backport?

Please avoid mentioning JDK-8297543, as it's already closed as a duplicate of JDK-8275731.

xiangzhai commented:
Hi @iklam 

Thanks for your kind response!

> The description of this PR is not clear.

I updated the description.  Please review it again.

Thanks,
Leslie Zhai

iklam commented:
> Hi @iklam
> 
> Thanks for your kind response!
> 
> > The description of this PR is not clear.
> 
> I updated the description. Please review it again.
> 
> > Are you backporting JDK-8267189 because it's required by [JDK-8275731](https://bugs.openjdk.org/browse/JDK-8275731) "CDS archived enums objects are recreated at runtime", which is the ultimate issue that you want to backport?
> 
> Yes.
> 
> Thanks, Leslie Zhai

I think the wording of this PR should be changed to something like this:

* Backporting JDK-8267189 as a prerequisite for JDK-8275731 (CDS archived enums objects are recreated at runtime)

You don't need to mention the other issues here. Instead, those issue should be listed in JDK-8275731:

* Add label jdk17u-fix-reques to JDK-8275731
* Add a comment in JDK-8275731, with something like
  * In order to backport this issue to jdk 17u, the following issues must be backported first, in the order of
    1.  JDK-8267189: Remove duplicated unregistered classes from dynamic archive
     2. JDK-XXXXXXX: xxxxxx
     3. JDK-XXXXXXX: xxxxxx
     4. JDK-XXXXXXX: xxxxxx

You should reach an agreement with the jdk17u maintainers first. I would suggest sending mail to the jdk-updates-dev mailing list. See https://mail.openjdk.org/pipermail/jdk-updates-dev/

You listed JDK-8270489 (Support archived heap objects in EpsilonGC) as a prerequisite, but that's a very large change and seems unrelated to the enum problem. More discussion is needed on jdk-updates-dev to decide which set of issues should be backported.

xiangzhai commented:
Hi @iklam 

Thanks for your teaching!

> I think the wording of this PR should be changed to something like this:

Updated.

> You don't need to mention the other issues here. Instead, those issue should be listed in JDK-8275731

Added.

> You should reach an agreement with the jdk17u maintainers first. I would suggest sending mail to the jdk-updates-dev mailing list.

May I just follow the [existing thread](https://mail.openjdk.org/pipermail/jdk-updates-dev/2022-December/019077.html) or create a new thread?

> You listed JDK-8270489 (Support archived heap objects in EpsilonGC) as a prerequisite, but that's a very large change and seems unrelated to the enum problem

Yes, there is a very large change, and owing to it requires `is_fully_available`:

The first commit about `HeapShared::is_fully_available` is for JDK-8270489: Support archived heap objects in EpsilonGC.
And the latest commit about `ArchiveHeapLoader::is_fully_available` is for JDK-8293293: Move archive heap loading code out of heapShared.cpp.

> More discussion is needed on jdk-updates-dev to decide which set of issues should be backported.

Yes!

Thanks,
Leslie Zhai

GoeLin commented:
@iklam 
Hi Ioi, thanks for advising Leslie!

GoeLin commented:
@xiangzha Hi Leslie,

please have a look at https://wiki.openjdk.org/display/JDKUpdates/How+to+contribute+or+backport+a+fix

You really should not integrate before you have a fix-yes label!

Also, please edit the comment in the JBS issue to contain the required information:
 * Why do you backport the change?
 * What is the risk of doing so?  E.g., will existing archives still work?
 * How you tested this
 * Only the fact whether it is clean or required edits is of interest in the JBS issue.  How you adapted the change goes into the PR as it is addressed at the reivewers.
 Fix this for all your backports please. 
Best, Goetz.

xiangzhai commented:
Hi @GoeLin 

Thanks for your teaching!

> You really should not integrate before you have a fix-yes label!

I see.

> Also, please edit the comment in the JBS issue to contain the required information:

Updated my comment in the JDK-8267189, please review it.

Thanks,
Leslie Zhai
